datajunction-server/datajunction_server/api/notifications.py

```diff
@@ -46,17 +46,35 @@ async def subscribe(
     session: AsyncSession = Depends(get_session),
     current_user: User = Depends(get_and_update_current_user),
 ) -> JSONResponse:
-    """Subscribes to notificaitons by upserting a notification preference"""
-    session.add(
-        NotificationPreference(
-            entity_type=entity_type,
-            entity_name=entity_name,
-            activity_types=activity_types,
-            alert_types=alert_types,
-            user=current_user,
+    """
+    Subscribes to notifications by upserting a notification preference.
+    If one exists, update it. Otherwise, create a new one.
+    """
+    result = await session.execute(
+        select(NotificationPreference).where(
+            NotificationPreference.entity_type == entity_type,
+            NotificationPreference.entity_name == entity_name,
+            NotificationPreference.user_id == current_user.id,
         ),
     )
+    existing = result.scalars().first()
+
+    if existing:
+        existing.activity_types = activity_types
+        existing.alert_types = alert_types
+    else:
+        session.add(
+            NotificationPreference(
+                entity_type=entity_type,
+                entity_name=entity_name,
+                activity_types=activity_types,
+                alert_types=alert_types,
+                user=current_user,
+            ),
+        )
```

Comment:
sqlalchemy has this [on_conflict_do_update](https://docs.sqlalchemy.org/en/20/dialects/postgresql.html#sqlalchemy.dialects.postgresql.Insert.on_conflict_do_update), would that work here? I've never used this method in sqlalchemy but from what I'm reading it converts to an `on conflict` query to postgres which I've used a lot and works great to upsert in a single query.